File: backend/src/models/Project.js
const { getOne, getAll, insert, update, remove, count } = require('../config/database');

class Project {
  static findById(id) {
    const project = getOne('projects', p => p.id === id);
    if (!project) return null;

    const owner = getOne('users', u => u.id === project.owner_id);
    const taskCount = count('tasks', t => t.project_id === id);
    const completedCount = count('tasks', t => t.project_id === id && t.status === 'done');

    return {
      ...project,
      owner_name: owner?.name || null,
      task_count: taskCount,
      completed_count: completedCount
    };
  }

  static create({ name, description, owner_id, status = 'active' }) {
    const project = insert('projects', { name, description, owner_id, status });
    return this.findById(project.id);
  }

  static updateById(id, fields) {
    const updated = update('projects', id, fields);
    if (!updated) return null;
    return this.findById(id);
  }

  static delete(id) {
    // Also delete associated tasks
    const tasks = getAll('tasks', t => t.project_id === id);
    tasks.forEach(t => remove('tasks', t.id));
    remove('projects', id);
  }

  static listByOwner(owner_id) {
    return getAll('projects', p => p.owner_id === owner_id)
      .map(p => {
        const taskCount = count('tasks', t => t.project_id === p.id);
        const completedCount = count('tasks', t => t.project_id === p.id && t.status === 'done');
        const owner = getOne('users', u => u.id === p.owner_id);
        return {
          ...p,
          owner_name: owner?.name || null,
          task_count: taskCount,
          completed_count: completedCount
        };
      })
      .sort((a, b) => new Date(b.updated_at) - new Date(a.updated_at));
  }

  static listAll() {
    return getAll('projects')
      .map(p => {
        const taskCount = count('tasks', t => t.project_id === p.id);
        const completedCount = count('tasks', t => t.project_id === p.id && t.status === 'done');
        const owner = getOne('users', u => u.id === p.owner_id);
        return {
          ...p,
          owner_name: owner?.name || null,
          task_count: taskCount,
          completed_count: completedCount
        };
      })
      .sort((a, b) => new Date(b.updated_at) - new Date(a.updated_at));
  }
}

module.exports = Project;

File: frontend/lib/models/project.dart
class Project {
  final int id;
  final String name;
  final String? description;
  final String status;
  final int ownerId;
  final String? ownerName;
  final int taskCount;
  final int completedCount;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Project({
    required this.id,
    required this.name,
    this.description,
    this.status = 'active',
    required this.ownerId,
    this.ownerName,
    this.taskCount = 0,
    this.completedCount = 0,
    this.createdAt,
    this.updatedAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      status: json['status'] ?? 'active',
      ownerId: json['owner_id'],
      ownerName: json['owner_name'],
      taskCount: json['task_count'] ?? 0,
      completedCount: json['completed_count'] ?? 0,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'description': description,
    'status': status,
  };

  double get progress => taskCount > 0 ? completedCount / taskCount : 0;
}
